run: local submitter fails (due mismatch of `parallel`?) but we fail to detect that and crash due to missing ref

Originally observed in https://github.com/ReproNim/reproman/issues/511#issuecomment-634914499 on smaug:

```
2020-05-27 16:07:27,208 [ERROR] CommandError: command 'git cat-file -p refs/reproman/20200527-160657-0490:.reproman/jobs/local/20200527-160657-0490/status.0' failed with exitcode 128
| Failed to run 'git cat-file -p refs/reproman/20200527-160657-0490:.reproman/jobs/local/20200527-160657-0490/status.0' under '/home/yoh/.tmp/rm-tfbYoMv/HOME/.reproman/run-root/940832c2-a055-11ea-a8dd-002590f97d84'. Exit code=128. out= err=fatal: Not a valid object name refs/reproman/20200527-160657-0490:.reproman/jobs/local/20200527-160657-0490/status.0
|  [orchestrators.py:_execute_in_wdir:641] (OrchestratorError)
```

and full stack would be (from a different run):
```
(Pdb) bt
  /home/yoh/proj/repronim/reproman-master/venvs/dev3/bin/reproman(11)<module>()
-> load_entry_point('reproman', 'console_scripts', 'reproman')()
  /home/yoh/proj/repronim/reproman-master/reproman/cmdline/main.py(250)main()
-> ret = cmdlineargs.func(cmdlineargs)
  /home/yoh/proj/repronim/reproman-master/reproman/interface/base.py(287)call_from_parser()
-> return cls.__call__(**kwargs)
  /home/yoh/proj/repronim/reproman-master/reproman/interface/run.py(420)__call__()
-> orc.follow()
  /home/yoh/proj/repronim/reproman-master/reproman/support/jobs/orchestrators.py(367)follow()
-> .format(self.jobid, self.status, self.working_directory))
  /home/yoh/proj/repronim/reproman-master/reproman/support/jobs/orchestrators.py(637)status()
-> self.working_directory)))
> /home/yoh/proj/repronim/reproman-master/reproman/support/jobs/orchestrators.py(617)_execute_in_wdir()
-> str(exc) if err_msg is None else err_msg)
```
, while working locally.

<details>
<summary>It might lead to by problem running parallel</summary> 

```shell
(git)smaug:…im/reproman-master/docs/usecases/bids-fmriprep-workflow-NP/out1[master]git
$> cat /home/yoh/.reproman/run-root/83d02dde-a0db-11ea-8df0-002590f97d84/.reproman/jobs/local/20200528-083028-15fd/stderr
Unknown option: c

$> cat /home/yoh/.reproman/run-root/83d02dde-a0db-11ea-8df0-002590f97d84/.reproman/jobs/local/20200528-083028-15fd/stdout 
Usage:

parallel [options] [command [arguments]] < list_of_arguments
parallel [options] [command [arguments]] (::: arguments|:::: argfile(s))...
cat ... | parallel --pipe [options] [command [arguments]]

-j n            Run n jobs in parallel
-k              Keep same order
-X              Multiple arguments with context replace
--colsep regexp Split input on regexp for positional replacements
{} {.} {/} {/.} {#} {%} {= perl code =} Replacement strings
{3} {3.} {3/} {3/.} {=3 perl code =}    Positional replacement strings
With --plus:    {} = {+/}/{/} = {.}.{+.} = {+/}/{/.}.{+.} = {..}.{+..} =
                {+/}/{/..}.{+..} = {...}.{+...} = {+/}/{/...}.{+...}

-S sshlogin     Example: foo@server.example.com
--slf ..        Use ~/.parallel/sshloginfile as the list of sshlogins
--trc {}.bar    Shorthand for --transfer --return {}.bar --cleanup
--onall         Run the given command with argument on all sshlogins
--nonall        Run the given command with no arguments on all sshlogins

--pipe          Split stdin (standard input) to multiple jobs.
--recend str    Record end separator for --pipe.
--recstart str  Record start separator for --pipe.

See 'man parallel' for details

Academic tradition requires you to cite works you base your article on.
When using programs that use GNU Parallel to process data for publication
please cite:

  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,
  ;login: The USENIX Magazine, February 2011:42-47.
```
</details>
and somehow then not having those references setup.

Note that above parallel is GNU's parallel (from `parallel` package on debian).  Locally (where execution worked) I had `parallel` from moreutils.   

So I guess the issue is multifold

- need to detect presence of the expected parallel (I think it should be the GNU one) and fail informatively if it is different; or support both (not sure if easy/good idea yet)
- capture the failed local submitter "execution"
- make that `.status` to raise a dedicated exception to indicate that job tracking is "errored" so `--follow` would just fail and `jobs` would report some ERROR

--- a/.travis.yml
+++ b/.travis.yml
@@ -90,7 +90,7 @@ install:
   - sudo sed -i -e 's/^Defaults.*secure_path.*$//' /etc/sudoers
   # for SVN tests (SVNRepoShim._ls_files_command())
   - sudo apt-get install sqlite3
-  - sudo apt-get install moreutils  # for concurrent jobs with local orchestrator
+  - sudo apt-get install parallel  # for concurrent jobs with local orchestrator
   - travis_retry sudo eatmydata apt-get install singularity-container
   - if [ ! -z "${INSTALL_DATALAD:-}" ]; then tools/ci/install_datalad; fi
   - if [ ! -z "${INSTALL_CONDOR:-}" ]; then tools/ci/install_condor; fi

--- a/reproman/support/jobs/job_templates/submission/local.template
+++ b/reproman/support/jobs/job_templates/submission/local.template
@@ -10,7 +10,7 @@ then
 else
     if test -z $(which parallel)
     then
-       echo "parallel (moreutils) is required to run concurrent jobs locally" >&2
+       echo "GNU parallel is required to run concurrent jobs locally" >&2
        exit 1
     fi
 
@@ -19,13 +19,16 @@ else
     metadir_rel={{ shlex_quote(_meta_directory_rel) }}
     workdir={{ shlex_quote(working_directory) }}
 
+    # Note: Orchestrator.submit() notifies the caller about the expectation to
+    # cite GNU Parallel.
     cd "$workdir"
-    parallel \
-        -i \
-        sh -c \
-        "$metadir_rel/runscript {} 1>$metadir_rel/stdout.{} 2>$metadir_rel/stderr.{}" \
-        -- $(seq 0 $(($num_subjobs - 1))) \
-        1>"$metadir_rel/stdout" 2>"$metadir_rel/stderr" &
+    seq 0 $(($num_subjobs - 1)) |
+        parallel \
+            --will-cite \
+            -q \
+            sh -c \
+            "$metadir_rel/runscript {} 1>$metadir_rel/stdout.{} 2>$metadir_rel/stderr.{}" \
+            1>"$metadir_rel/stdout" 2>"$metadir_rel/stderr" &
 fi
 
 RUNSCRIPT_PID=$!

--- a/reproman/support/jobs/job_templates/submission/local.template
+++ b/reproman/support/jobs/job_templates/submission/local.template
@@ -10,7 +10,7 @@ then
 else
     if test -z $(which parallel)
     then
-       echo "parallel (moreutils) is required to concurrent jobs locally" >&2
+       echo "parallel (moreutils) is required to run concurrent jobs locally" >&2
        exit 1
     fi
 

--- a/reproman/support/jobs/job_templates/submission/local.template
+++ b/reproman/support/jobs/job_templates/submission/local.template
@@ -8,9 +8,17 @@ if test $num_subjobs -eq 1
 then
     "$metadir/runscript" 0 1>"$metadir/stdout.0" 2>"$metadir/stderr.0" &
 else
-    if test -z $(which parallel)
+    dryerr=0
+    echo '' | parallel --will-cite echo >/dev/null 2>&1 || dryerr=$?
+
+    if test $dryerr -ne 0
     then
-       echo "GNU parallel is required to run concurrent jobs locally" >&2
+       err="GNU parallel is required to run concurrent jobs locally"
+       if test $dryerr -ne 127
+       then
+           err="$err. An incompatible 'parallel' program (likely from moreutils) is installed."
+       fi
+       echo "$err" >&2
        exit 1
     fi
 

--- a/reproman/support/jobs/orchestrators.py
+++ b/reproman/support/jobs/orchestrators.py
@@ -111,11 +111,17 @@ class Orchestrator(object, metaclass=abc.ABCMeta):
 
         self.template = None
 
-    def _find_root(self):
+    @property
+    @cached_property
+    def home(self):
+        "$HOME directory on resource."
         home = self.session.query_envvars().get("HOME")
         if not home:
             raise OrchestratorError("Could not determine $HOME on remote")
-        root_directory = op.join(home, ".reproman", "run-root")
+        return home
+
+    def _find_root(self):
+        root_directory = op.join(self.home, ".reproman", "run-root")
         lgr.info("No root directory supplied for %s; using '%s'",
                  self.resource.name, root_directory)
         if not op.isabs(root_directory):

--- a/reproman/support/jobs/orchestrators.py
+++ b/reproman/support/jobs/orchestrators.py
@@ -630,13 +630,20 @@ class PrepareRemoteDataladMixin(object):
         if status == "unknown":
             # The local tree might be different because of another just. Check
             # the ref for the status.
-            status_from_ref = self._execute_in_wdir(
-                "git cat-file -p {}:{}"
-                .format(self.job_refname,
-                        # FIXME: How to handle subjobs?
-                        op.relpath(op.join(self.meta_directory, "status.0"),
-                                   self.working_directory)))
-            status = status_from_ref.strip() or status
+            try:
+                status_from_ref = self._execute_in_wdir(
+                    "git cat-file -p {}:{}"
+                    .format(self.job_refname,
+                            # FIXME: How to handle subjobs?
+                            op.relpath(op.join(self.meta_directory, "status.0"),
+                                       self.working_directory)))
+            except OrchestratorError as exc:
+                # Most likely the ref was never created because the runscript
+                # failed. Let follow() signal the error.
+                lgr.debug("Failed to get status from %s tree: %s",
+                          self.job_refname, exc_str(exc))
+            else:
+                status = status_from_ref.strip() or status
         return status
 
     def get_failed_subjobs(self):

--- a/reproman/support/jobs/orchestrators.py
+++ b/reproman/support/jobs/orchestrators.py
@@ -363,8 +363,9 @@ class Orchestrator(object, metaclass=abc.ABCMeta):
         # post-processing. Make sure it looks like it passed.
         if not self.has_completed:
             raise OrchestratorError(
-                "Post-processing failed for {} [status: {}] ({})"
-                .format(self.jobid, self.status, self.working_directory))
+                "Runscript handling failed for {} [status: {}]\n"
+                "Check error logs in {}"
+                .format(self.jobid, self.status, self.meta_directory))
 
     def _get_io_set(self, which, subjobs):
         spec = self.job_spec

--- a/reproman/support/jobs/orchestrators.py
+++ b/reproman/support/jobs/orchestrators.py
@@ -239,11 +239,20 @@ class Orchestrator(object, metaclass=abc.ABCMeta):
     def submit(self):
         """Submit the job with `submitter`.
         """
+        njobs = len(self.job_spec["_command_array"])
+        if njobs > 1 and self.submitter.name == "local":
+            will_cite = op.join(self.home, ".parallel", "will-cite")
+            if not self.session.exists(will_cite):
+                lgr.info(
+                    "Submitter will use GNU Parallel.\n"
+                    "Its author asks that you cite it. "
+                    "Please run `parallel --citation` on the resource '%s'",
+                    self.resource.name)
         lgr.info("Submitting %s", self.jobid)
         templ = Template(
             **dict(self.job_spec,
                    _jobid=self.jobid,
-                   _num_subjobs=len(self.job_spec["_command_array"]),
+                   _num_subjobs=njobs,
                    root_directory=self.root_directory,
                    working_directory=self.working_directory,
                    _meta_directory=self.meta_directory,

